File: docs/manual/guide/cluster/code/docs/home/persistence/BlogServiceImpl3.java
package docs.home.persistence;

import scala.PartialFunction;

import java.util.Optional;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;
import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraReadSide;
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.cassandra.CassandraSession;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;

import akka.japi.pf.PFBuilder;
import akka.stream.javadsl.Source;

public class BlogServiceImpl3 implements BlogService2 {

  private final PersistentEntityRegistry persistentEntityRegistry;
  private final CassandraSession cassandraSession;

  //#register-event-processor
  @Inject
  public BlogServiceImpl3(
      PersistentEntityRegistry persistentEntityRegistry,
      CassandraSession cassandraSession,
      CassandraReadSide cassandraReadSide) {

    this.persistentEntityRegistry = persistentEntityRegistry;
    this.cassandraSession = cassandraSession;

    cassandraReadSide.register(BlogEventProcessor.class);
  }
  //#register-event-processor

  @Override
  public ServiceCall<NotUsed, Source<PostSummary, ?>> getPostSummaries() {
    return request -> {
      Source<PostSummary, ?> summaries = cassandraSession.select(
          "SELECT id, title FROM postsummary;").map(row ->
            PostSummary.of(row.getString("id"), row.getString("title")));
      return CompletableFuture.completedFuture(summaries);
    };
  }

  //#event-stream
  public ServiceCall<NotUsed, Source<PostSummary, ?>> newPosts() {
    final PartialFunction<BlogEvent, PostSummary> collectFunction =
        new PFBuilder<BlogEvent, PostSummary>()
        .match(PostAdded.class, evt ->
           PostSummary.of(evt.getPostId(), evt.getContent().getTitle()))
        .build();

    return request -> {
      Source<PostSummary, ?> stream = persistentEntityRegistry
        .eventStream(BlogEventTag.INSTANCE, Optional.empty())
          .map(pair -> pair.first()).collect(collectFunction);
      return CompletableFuture.completedFuture(stream);
    };
  }
  //#event-stream

}

File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/BlogServiceImpl3.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

import com.lightbend.lagom.scaladsl.persistence.ReadSide

import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry
import akka.NotUsed
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.api.Service
import akka.stream.scaladsl.Source
import akka.persistence.query.NoOffset
import com.lightbend.lagom.scaladsl.persistence.EventStreamElement
import scala.concurrent.Future

trait BlogServiceImpl3 {
  trait BlogService extends Service {
    def newPosts(): ServiceCall[NotUsed, Source[PostSummary, _]]

    override def descriptor = ???
  }

  //#register-event-processor
  class BlogServiceImpl(persistentEntityRegistry: PersistentEntityRegistry, readSide: ReadSide, myDatabase: MyDatabase)
      extends BlogService {

    readSide.register[BlogEvent](new BlogEventProcessor(myDatabase))
    //#register-event-processor

    //#event-stream
    override def newPosts(): ServiceCall[NotUsed, Source[PostSummary, _]] =
      ServiceCall { request =>
        val response: Source[PostSummary, NotUsed] =
          persistentEntityRegistry
            .eventStream(BlogEvent.Tag.forEntityId(""), NoOffset)
            .collect {
              case EventStreamElement(entityId, event: PostAdded, offset) =>
                PostSummary(event.postId, event.content.title)
            }
        Future.successful(response)
      }
    //#event-stream
  }
}
